server/src/main/java/io/druid/client/cache/MemcacheClientPool.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.client.cache;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import io.druid.collections.ResourceHolder;
+import io.druid.java.util.common.logger.Logger;
+import net.spy.memcached.MemcachedClientIF;
+import sun.misc.Cleaner;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Simple load balancing pool that always returns the least used {@link MemcachedClientIF}.
+ *
+ * A client's usage is incremented every time one gets requested from the pool
+ * and is decremented every time close is called on the holder.
+ *
+ * The pool eagerly instantiates all the clients in the pool when created,
+ * using the given supplier.
+ */
+final class MemcacheClientPool implements Supplier<ResourceHolder<MemcachedClientIF>>
+{
+  private static final Logger log = new Logger(MemcacheClientPool.class);
+
+  /**
+   * The number of memcached connections is not expected to be small (<= 8), so it's easier to find the least used
+   * connection using a linear search over a simple array, than fiddling with PriorityBlockingQueue. This also allows
+   * to reduce locking.
+   */
+  private final CountingHolder[] connections;
```

Comment:
this is the only major difference I see , older class used PriorityBlockingQueue instead to have logN get operations. I guess, size of this pool is usually very small so that difference shouldn't matter